### ISSUE TITLE
Denying fe.protection.mobspawn.forced.* does not prevent spawning.

### ISSUE THREAD

**Issue by DashFury** [ForgeEssentials/ForgeEssentials#1990]
As of version 1.4.4.1146 (for Minecraft 1.7.10, with Forge 10.13.4.1558), denying `fe.protection.mobspawn.forced.*` does not prevent mobs from spawning from mob spawners, despite being, according to [the wiki](https://github.com/ForgeEssentials/ForgeEssentials/wiki/Protection-%28world-guard%29#mob-permissions), the intended behaviour.

I don't know if this is linked or not, but denying `fe.protection.mobspawn.forced.PigZombie` or `fe.protection.mobspawn.natural.PigZombie` will not stop Zombie Pigmen from spawning from ender portal frames in the Overworld.

Denying `fe.protection.mobspawn.natural.*` works as intended.

The commands I used, to be sure I'm not mistaken:
```
/p global zone WORLD_10 deny fe.protection.mobspawn.forced.*
/p global zone WORLD_10 deny fe.protection.mobspawn.natural.*
```
This is in a multiworld. I'm pretty sure I didn't make any mistake since the natural spawn, as stated above, is effectively stopped.

For the pigmen, I used:
```
/p global zone villagespawn deny fe.protection.mobspawn.forced.PigZombie
/p global zone villagespawn deny fe.protection.mobspawn.natural.PigZombie
```
The ender portal is set in the defined zone "villagespawn". I denied both since I wasn't getting the expected result with only one, but they're still spawning anyway.

**Comment by Entoarox** [comment]
Same happened for me with pigmen & the portal, so consider it verified.

**Comment by damonmensch** [comment]
@Entoarox pigmen don't spawn from the portal. They travel THROUGH it, just like any normal player would.

**Comment by Entoarox** [comment]
@damonmensch Since the other side of the portal is surrounded by a quite large zone to prevent the exact same thing, it would still qualify as a issue.
Further, the nether-side is not chunkloaded, yet pigmen can still spawn, so no, this is NOT pigmen travelling.

**Comment by damonmensch** [comment]
Even if the other end isn't loaded, minecraft still generates the pigmen as if it was a pigmen traveling through the gate. The portal blocks themselves NEVER run a spawn command, only a teleport command. However when the area isn't actually loaded on the other side of the portal, it tends to act like a spawner, but is really only runnning teleport commands on entities. However when the chunk/world isn't loaded in memory, chances are your protections on the nether side aren't being loaded either.

**Comment by Entoarox** [comment]
They are still being spawned, and thus, FE's permission system should be able to block that from happening.
And it doesnt matter if the protections arent active if the chunk isnt loaded, cause if the chunk isnt loaded mobs are not loaded either, so they cant be spawning on the nether side either way!

**Comment by DashFury** [comment]
> @Entoarox pigmen don't spawn from the portal. They travel through it, just like any normal player would.

No, they don't come from the other side. Check the Minecraft wiki, pigmen do spawn in the Overworld from portal frames. Sure, they _can_ travel like normal players, but when the Nether isn't loaded and you still see pigmen appearing in portals, it's because they spawn there.

> The portal blocks themselves NEVER run a spawn command, only a teleport command.
next

Absolutely not. The game does not load some secret invisible location in which it spawns pigmen only to teleport them on portal frames. It just directly spawns them in the portal. Why would it ever be so complicated?

> However when the chunk/world isn't loaded in memory, chances are your protections on the nether side aren't being loaded either.

ForgeEssentials permissions are loaded all the time. They're not saved with the chunks they affect.

And just to make things clear, I tested this for hours. I blocked every valid spawn location around the portal on the nether side. I barred it, then dig a hole in front of it instead, so no entity could cross. I made it physically impossible for any entity to cross the portal in any direction, and yet I would still find pigmen in the Overworld.

And even then, just to be extremely sure, supposing they could still find a way to cross, and even though only overworld portals have the ability to spawn pigmen, I denied the spawn on the Nether one too. And yet I would still find them in the Overworld.

**Comment by olee** [comment]
Minecraft by default keeps the area around the portal in the Nether always loaded afaik.
That's why Pigmen can travel through them.
Could you try surrounding the portal on the Nether side with walls / doors and check if the issue really still happens?
If they are really spawned from the portal, I think it would use the same mechanics as mob spawners which would let FE's permissions work.

**Comment by Entoarox** [comment]
Having a spawn denial zone around the nether side didnt work in preventing the overworld-side spawns either.

And @DashFury already mention he tested it out by preventing any spawns on the nether-side.

**Comment by damonmensch** [comment]
As I said before, that's because minecraft doesn't run the spawner tick, it creates an entity and then teleports it. It may seem illogical and ass backwards, but that's how it does it. I've run in to this problem before.

Best work around would probably either have the spawn protection also despawn mobs that enter the area, or create a separate perm for it.

**Comment by DashFury** [comment]
It doesn't explain why denying `fe.protection.mobspawn.forced.*` doesn't block mob spawners, though.

**Comment by DashFury** [comment]
I couldn't do this earlier as I was quite busy, but now that I am returning to this thread, I took the time to decompile Minecraft (1.7.10, although that part of the code probably didn't change in quite a while nor in the recent versions) and here is what I found in `BlockPortal.java`:
> Entity var7 = ItemMonsterPlacer.spawnCreature(p_149674_1_, 57, (double)p_149674_2_ + 0.5D, (double)var6 + 1.1D, (double)p_149674_4_ + 0.5D);

This is, without possible doubt, a spawn happening right at the coordinates above the first solid block found under the ticking portal block. It calls `ItemMonsterPlacer.spawnCreature` which I am sure bypasses ForgeEssentials' protections. `ItemMonsterPlacer` is usually used by the eggs, but apparently by portal blocks too.

There's also some lines to make the entity behave like it just travelled through a portal, like the portal cooldown, the `inPortal` flag, that kind of things, but it definitively didn't got teleported from anywhere. It spawned right in the portal, as if someone used an egg there.

So, as @damonmensch says, it isn't spawned by the spawner tick, which means `fe.protection.mobspawn.natural.*` should not affect it. Alright. But there's no reason for `fe.protection.mobspawn.forced.*` to ignore it. And as stated before, this second set of rules doesn't even affect mob spawners, so there's definitively a problem.

**Comment by luacs1998** [comment]
Thanks for the investigation.

I'll probably insert a hook there if needed to apply permissions.

**Comment by luacs1998** [comment]
I spoke to Lex, this is a Forge bug...

**Comment by spacebuilder2020** [comment]
@DashFury Has this issue been resolved?

**Comment by Zethalion** [comment]
Not resolved
